src/ZODB/Connection.py

```diff
@@ -866,109 +877,42 @@ def setstate(self, obj):
+                # updated since this transaction, and been packed
+                # away, cuz the tests are mean. The best we can do is
+                # raise a ReadConflictError and try again
                 raise ReadConflictError()
 
-            if (oid in self._invalidated):
-                self._load_before_or_conflict(obj)
-                return
-
-            p, serial = self._storage.load(oid, '')
+            p, serial, _ = data
             self._load_count += 1
 
-            self._inv_lock.acquire()
-            try:
-                invalid = oid in self._invalidated
-            finally:
-                self._inv_lock.release()
-
-            if invalid:
-                self._load_before_or_conflict(obj)
-                return
-
-        self._reader.setGhostState(obj, p)
-        obj._p_serial = serial
-        self._cache.update_object_size_estimation(oid, len(p))
-        obj._p_estimated_size = len(p)
-
-        # Blob support
-        if isinstance(obj, Blob):
-            obj._p_blob_uncommitted = None
-            obj._p_blob_committed = self._storage.loadBlob(oid, serial)
-
-    def _load_before_or_conflict(self, obj):
-        """Load non-current state for obj or raise ReadConflictError."""
-        if not self._setstate_noncurrent(obj):
-            self._register(obj)
-            self._conflicts[obj._p_oid] = True
-            raise ReadConflictError(object=obj)
+            self._reader.setGhostState(obj, p)
+            obj._p_serial = serial
+            self._cache.update_object_size_estimation(oid, len(p))
+            obj._p_estimated_size = len(p)
 
-    def _setstate_noncurrent(self, obj):
-        """Set state using non-current data.
+            # Blob support
+            if isinstance(obj, Blob):
+                obj._p_blob_uncommitted = None
+                obj._p_blob_committed = self._storage.loadBlob(oid, serial)
 
-        Return True if state was available, False if not.
-        """
-        try:
-            # Load data that was current before the commit at txn_time.
-            t = self._storage.loadBefore(obj._p_oid, self._txn_time)
-        except KeyError:
-            return False
-        if t is None:
-            return False
-        data, start, end = t
-        # The non-current transaction must have been written before
-        # txn_time.  It must be current at txn_time, but could have
-        # been modified at txn_time.
-
-        assert start < self._txn_time, (u64(start), u64(self._txn_time))
-        assert end is not None
-        assert self._txn_time <= end, (u64(self._txn_time), u64(end))
-        self._reader.setGhostState(obj, data)
-        obj._p_serial = start
-
-        # MVCC Blob support
-        if isinstance(obj, Blob):
-            obj._p_blob_uncommitted = None
-            obj._p_blob_committed = self._storage.loadBlob(obj._p_oid, start)
-
-        return True
+        except ConflictError:
+            raise
+        except:
```

Comment:
As a rule, I try not to make changes unrelated to what I'm doing to try to limit the sizes of PRs.